### DISCUSSION TITLE
How to override format: html: cap-location: margin

### DISCUSSION THREAD

**Discussion by harrelfe** [quarto-dev/quarto-cli#1172] [Q&A]
I specify cap-location: margin in my Quarto book yaml.  For the few figures that have multiple panels this causes the figures to no longer be rendered in a matrix but rather as larger individual plots.  Is there a way to override cap-location for one code chunk?  Here's some test code:

---
title: "Test"
format:
  html:
    self-contained: true
    cap-location: margin
---

```{r}
#| label: fig-test
#| fig.width: 8
#| fig.height: 8
#| fig.scap: 'Four examples of residual plots'
#| fig.cap: 'Using residuals to check some of the assumptions of the simple linear regression model.'
#| fig-subcap:
#|   - "Non-constant $\\sigma^2$, which might call for transforming $y$"
#|   - "Constant $\\sigma^2$ but systemic failure of the linearity assumption"
#|   - "Ideal situation of white noise (no trend, constant variance)"
#|   - "$q-q$ plot demonstrating approximate normality of residuals, for a sample of size $n=50$.  Horizontal reference lines are at zero, which is by definition the mean of all residuals."
#| layout-ncol: 2
plot(rnorm(20)); plot(runif(20)); hist(rnorm(100)); hist(rexp(100))
```

**Comment by mcanouil** [comment]
You can use the option as cell chunk option.

````
---
title: "A Quarto Page Layout Example"
format:
  html: 
    self-contained: true
---

# Tables

```{r}
#| tbl-cap-location: bottom
knitr::kable(
  x = mtcars[1:6, 1:6],
  caption = 'A subset of mtcars.'
)
```

```{r}
#| tbl-cap: "A subset of iris."
#| tbl-cap-location: margin
knitr::kable(head(iris[,c(1,2,3,4)], 15))
```


+---------------+---------------+--------------------+
| Fruit         | Price         | Advantages         |
+===============+===============+====================+
| Bananas       | $1.34         | - built-in wrapper |
|               |               | - bright color     |
+---------------+---------------+--------------------+
| Oranges       | $2.10         | - cures scurvy     |
|               |               | - tasty            |
+---------------+---------------+--------------------+

: Sample grid table.

# Figures

```{r}
#| fig.cap: "Plot 1 to 5"
#| fig-cap-location: bottom
plot(1:5, 1:5)
```

```{r}
#| fig-cap: "Density plot"
#| fig-cap-location: margin
plot(density(rnorm(100)))
```

````

**Reply by harrelfe** [reply]
Thanks for that.  When making a book and marginal captions work for 9/10ths of the figures, and the book has only one yaml for all chapters, it's more challenging to use that solution.

**Reply by mcanouil** [reply]
This solution is only to set/override locally the option, of course, doing that for all cell chunk is a bad idea.
But for now, it is not working as intended, the global option is not overwritten by the local ones.
See #1173 opened by @dragonstyle 
Once this is fixed, the above will answer fully your question.

**Comment by dragonstyle** [comment]
As @mcanouil noted, you can set the location per chunk, but right now it looks like you can't set the document globally to use the margin location and then override that on a per chunk basis. I'll try to take a quick look this weekend to see if that is something we could make work, as I think it would be pretty helpful for large documents...

**Reply by harrelfe** [reply]
How's it looking for a possible fix?  Thanks.

**Reply by jjallaire** [reply]
I don't think we are going to do this for v1.0 (fix is non-trivial) however we will 100% look at it for v1.1 (converting this thread to an issue to track the work).

**Reply by mcanouil** [reply]
@jjallaire @dragonstyle already created an issue for this, namely #1173

**Reply by harrelfe** [reply]
I thought I had a workaround by creating a `knitr` hook that at the start of each chunk would execute a `fig.cap.location` setting.  But alas the following doesn't work in a chunk header for caption positioning so it won't work either with `knitr::opts_chunk$set`:

```
{r fig.cap.location='...'}
```

**Comment by harrelfe** [comment]
My last reply is not very visable so I'd like to restate it here.  Is it possible to modify `Quarto` so that `knitr::opts_chunk$set(fig.cap.location=...)` works?  That way I could easily program changes in caption locations for whole chapters.

Or is this really the same task as processing the `yaml` in the chunk anyway, for which version 1.1 will do this?

**Comment by fredguth** [comment]
How do you override the cap location per chunk with an image file?  Most examples on the `Figures`  section of the documentation cover only when the figures are created in R.  That is ok, but there is still the case when the project has outside figures that we want to use. 

Maybe the solution is to use a R code that show the image, but I have never used R before.

**Comment by fredguth** [comment]
Just to illustrate, look what happens when I try to add a figure in the margin: https://share.getcloudapp.com/E0uZoodE

This is the code.
```md
:: {.column-margin #fig-ars_magna_disc label="fig-ars_magna_disc"}
![](imgs/ars_magna_disc.png)

Example of one of Lull’s Ars Magna’s paper discs.
:::

In the 13^th^ century, the Catalan philosopher Ramon Lull wanted to produce all statements the human mind can think. For this task, he developed *logic paper machines*, discs of paper filled with esoteric coloured diagrams that connected symbols representing statements. Unfortunately, according to @gardner:1959, in a modern reassessment of his work, *"it is impossible, perhaps, to avoid a strong sense of anticlimax"* [@gardner:1959]. With megalomaniac self-esteem that suggests psychosis, his delusional sense of importance is more characteristic of cult founders. On the bright side, his ideas and books exerted some magic appeal that helped them be rapidly disseminated through all Europe [@gardner:1959].

::: {.column-margin}
![](imgs/ars_magna_disc.png)

Example of one of Lull’s Ars Magna’s paper discs.
:::

````

I don't know how to add a label to the margin figure.  I also have tried:
```md
::: {.column-margin}
![](imgs/ars_magna_disc.png){#fig-ars_magna_disc label="fig-ars_magna_disc"}

Example of one of Lull’s Ars Magna’s paper discs.
:::
```

but didn't work.

**Comment by kjohnsen** [comment]
I am also interested in being able to override caption location on a (non-code) per-figure basis. I would expect there to be a CSS parameter for the div like
```
::: {#fig-example cap-location=margin}
![](image.jpg)

my caption
:::
```

Oh wait...I'm discovering the top-level YAML config isn't working for LaTeX...why is that? I tried fig-cap-location: margin and it only worked for my code-generated figure and not all div ones. All captions are in the margin as expected for HTML 

EDIT: I've discovered that in LaTeX, the div-style figures with captions don't respond to `fig-cap-location`, whereas Pandoc's 'implicit figures' (`![caption](img.png){#fig-myfig}`) do
